### PR TITLE
Implement IsAlreadyMirrored & cmd/is-falce-ebpf-probe-uploaded & add a temporary workflow to test

### PR DESCRIPTION
**Pull request by lemanie**
I have a few questions about implementing this ticket:
1. You mentioned I would need to use github actions on a feature branch instead of personal access token to test IsAlreadyMirrored, I'm not sure how that would work? Would we just add it in temporarily to test this ticket, then remove it during the next ticket where it gets used by the actual build action?
2. I don't understand how the //cmd part can possibly work without using 
a) a personal access token 
b) an extra arg for using the repo.IsAlreadyMirrored?
3. Even with a personal access token I'm getting a 401 with
`plz run //cmd/is-falco-ebpf-probe-uploaded -- --verbose --falco_version=0.28.1 --github_releases_token=$GRT amazonlinux2 4.14.200-155.322.amzn2
error="could not list releases: GET https://api.github.com/repos/thought-machine/falco-probes/releases: 401 Bad credentials []"` I'm not sure if this is expected or something I need to configure on my github account?
4. For IsAlreadyMirrored I added a returned error is that ok?

### CODE REVIEW

**Comment by VJftw** [comment]
> I have a few questions about implementing this ticket:
> 
> 1. You mentioned I would need to use github actions on a feature branch instead of personal access token to test IsAlreadyMirrored, I'm not sure how that would work? Would we just add it in temporarily to test this ticket, then remove it during the next ticket where it gets used by the actual build action?

Yeah, that's exactly what I was thinking :) though, you're right this isn't a nice thing to do. Ideally we'd be writing unit tests to validate this which would get picked up by the "Test" github workflow. I started trying to get this working here: https://github.com/thought-machine/falco-probes/blob/unit-test-releases/pkg/repository/ghreleases/ghreleases_test.go#L31 but haven't quite got it yet. So i'm happy for any workarounds we have just to demonstrate that it works (i.e. pasting command output, or a temporary workflow).

> 2. I don't understand how the //cmd part can possibly work without using
>    a) a personal access token
>    b) an extra arg for using the repo.IsAlreadyMirrored?

I'm not sure that I understand your question here; the `//cmd/is-falco-ebpf-probe-uploaded`'s job is to check if the probe has already been uploaded. This requires _a_ and is calling _b_. Ah right, sorry the command i've given you in your ticket doesn't detail the inputs. What you've done is 💯  what I was expecting 👍  i.e. `plz run //cmd/is-falco-ebpf-probe-uploaded -- --falco_version <falco_version> <operating_system> <kernel_package_name>`

> 3. Even with a personal access token I'm getting a 401 with
>    `plz run //cmd/is-falco-ebpf-probe-uploaded -- --verbose --falco_version=0.28.1 --github_releases_token=$GRT amazonlinux2 4.14.200-155.322.amzn2 error="could not list releases: GET https://api.github.com/repos/thought-machine/falco-probes/releases: 401 Bad credentials []"` I'm not sure if this is expected or something I need to configure on my github account?

Strange, it's working for me: 
```
# Create github personal token from https://github.com/settings/tokens with no scopes attached
$ vim .github_token
# paste github token into file (can use nano)
$ export GITHUB_TOKEN=$(cat .github_token)
$ plz run //cmd/is-falco-ebpf-probe-uploaded -- --verbose --falco_version=0.29.1 amazonlinux2 4.14.200-155.322.amzn2
...
11:05AM INF Identifying if falco ebpf probe uploaded for  driverVersion=17f5df52a7d9ed6bb12d3b1768460def8439936d probeName=falco_amazonlinux2_4.14.200-155.322.amzn2.x86_64_1.o
11:05AM INF Matching release found, now checking it's assets:  release=17f5df52a7d9ed6bb12d3b1768460def8439936d
11:05AM FTL Probe cannot be found: error="release/asset not found that matches 17f5df52a7d9ed6bb12d3b1768460def8439936d/falco_amazonlinux2_4.14.200-155.322.amzn2.x86_64_1.o"
```

> 4. For IsAlreadyMirrored I added a returned error is that ok?

Yup, that's okay! nice one 👍

**Comment by lemanie** [comment]
thanks @VJftw! I'll work on these changes now!

**Comment by lemanie** [comment]
Tested manually using:
`#   plz run //cmd/is-falco-ebpf-probe-uploaded -- --falco_version=0.29.1 amazonlinux2 4.14.200-155.322.amzn2 -> 1
#   plz run //cmd/is-falco-ebpf-probe-uploaded -- --falco_version=0.29.1 amazonlinux2 4.14.101-91.76.amzn2  -> 0 `

**Comment by lemanie** [comment]
Added in a workflow & accompanying build script:
- .github/workflows/test-is-falco-ebpf-probe-uploaded.yaml
- build/github/test-is-falco-ebpf-probe-uploaded
These are a bit quick and dirty but we plan to delete them after the next ticket is completed where we will actually use IsAlreadyMirrored in the existing build script.

**Comment by lemanie** [comment]
2 test cases are still failing when they're meant to be passing, so I'm looking into it.

**Comment by lemanie** [comment]
Added in pagination, tested locally using:
`./pleasew -p -v2 run //build/github/test-is-already-mirrored -- amazonlinux2     `

**Comment by lemanie** [comment]
@VJftw one last question, I left the verifying input parts to the cmd not IsAlreadyMirrored. I thought it would be duplicating code when IsAlreadyMirrored is used properly, do you agree?

File: pkg/repository/ghreleases/ghreleases.go [hunk 1]
```diff
@@ -82,9 +82,54 @@ func (ghr *GHReleases) PublishProbe(driverVersion string, probePath string) erro
 }
 
 // IsAlreadyMirrored implmements repository.Repository.IsAlreadyMirrored for GitHub Releases.
-func (ghr *GHReleases) IsAlreadyMirrored(driverVersion string, probeName string) bool {
-	// TODO: unimplimented
-	return false
+func (ghr *GHReleases) IsAlreadyMirrored(driverVersion string, probeName string) (bool, error) {
+	// use a mutex to ensure that this function is only called once at a time as it is not thread safe.
+	// i.e. a release that doesn't exist may result in multiple goroutines trying to create it at once.
```
Comment (VJftw) [inline pkg/repository/ghreleases/ghreleases.go:87]: I don't think you need a mutex here (and the "i.e" doesn't fit with what this function is doing) as you're only listing releases.

File: pkg/repository/ghreleases/ghreleases.go [hunk 2]
```diff
@@ -82,9 +82,54 @@ func (ghr *GHReleases) PublishProbe(driverVersion string, probePath string) erro
...
+	ghr.releasesMu.Lock()
+	defer ghr.releasesMu.Unlock()
+	ctx := context.Background()
+
+	// Retrieve the releases
```
Comment (VJftw) [inline pkg/repository/ghreleases/ghreleases.go:92]: There's a lot of nesting going on under here. Try thinking about how you could break this into smaller functions to reduce that nesting. e.g.
```go
release, err := ghr.getReleaseByName(driverVersion)
asset, err := ghr.getAssetFromReleaseByName(release, probeName)
```

File: pkg/repository/ghreleases/ghreleases.go [hunk 3]
```diff
@@ -82,9 +82,54 @@ func (ghr *GHReleases) PublishProbe(driverVersion string, probePath string) erro
...
+	currentReleasePage := 0
+	lastReleasePage := 0
```
Comment (VJftw) [inline pkg/repository/ghreleases/ghreleases.go:94]: I suggest following the same pattern in their example: https://pkg.go.dev/github.com/google/go-github/v37/github#hdr-Pagination
e.g. for iterating through releases, it'd be
```go
        opt := &github.ListOptions{PerPage: 100}
	for {
		releases, releasesResp, err := ghr.ghClient.Repositories.ListReleases(ctx, ghr.owner, ghr.repo, opt)
		if err != nil {
			return false, fmt.Errorf("could not list releases: %w", err)
		}
		if resp.NextPage == 0 {
			break
		}
		opt.Page = resp.NextPage
	}
```

**Review by VJftw — COMMENTED** [review]
nice one! my comments are just some general improvements but you've gone down the right tracks 👌

**Comment by VJftw** [comment]
> @VJftw one last question, I left the verifying input parts to the cmd not IsAlreadyMirrored. I thought it would be duplicating code when IsAlreadyMirrored is used properly, do you agree?

Yes, there's very little value to verifying inputs as there are no negative consequences for bad inputs in the tooling we're making here. i.e. doing `plz run //cmd/is-falco-ebpf-probe-uploaded -- amazonlinux2 as;djfhsladjfhj` will just return an error saying that the package could not be downloaded 👍

File: pkg/repository/ghreleases/ghreleases.go (hunk 3 above)
Comment (lemanie) [reply pkg/repository/ghreleases/ghreleases.go]: I'm now using the example pattern, thanks!

File: pkg/repository/ghreleases/ghreleases.go (hunk 2 above)
Comment (lemanie) [reply pkg/repository/ghreleases/ghreleases.go:88]: I've split these 2 out into their own functions as suggested.

File: pkg/repository/ghreleases/ghreleases.go (hunk 1 above)
Comment (lemanie) [reply pkg/repository/ghreleases/ghreleases.go]: I've removed the mutex and the comment.

File: cmd/is-falco-ebpf-probe-uploaded/main.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		log.Fatal().Err(err).Msg("Could not get operating system")
+	}
+
+	log.Info().Str("kernel_package", opts.Positional.KernelPackage).Msg("Verifying input")
+	kernelPackage, err := operatingSystem.GetKernelPackageByName(opts.Positional.KernelPackage)
+	if err != nil {
+		log.Fatal().Err(err).Msg("Could not get kernel package")
+	}
+	probeName := kernelPackage.ProbeName() + ".o"
+	if strings.Contains(probeName, operatingSystem.GetName()) == false {
```
[the remark this answers is not in the document]
Comment (lemanie) [reply cmd/is-falco-ebpf-probe-uploaded/main.go]: I've moved this to pkg instead.

File: pkg/repository/ghreleases/ghreleases.go [hunk 5]
```diff
@@ -82,9 +82,67 @@ func (ghr *GHReleases) PublishProbe(driverVersion string, probePath string) erro
...
+	}
+
+	log.Info().Str("using", *asset.BrowserDownloadURL).Msg("Probe is uploaded and available")
+	return true, nil
+}
+
+// getAssetFromReleaseByName uses the github API to identify whether the desired probe is an asset of the given release
+func (ghr *GHReleases) getAssetFromReleaseByName(ctx context.Context, release *github.RepositoryRelease, probeName string) (*github.ReleaseAsset, error) {
+	// Retrieve the release's assets
+	opt := &github.ListOptions{PerPage: 20}
```
Comment (VJftw) [inline pkg/repository/ghreleases/ghreleases.go:105]: nit: before you merge, don't forget to update this to `100`

File: pkg/repository/ghreleases/ghreleases.go [hunk 6]
```diff
@@ -82,9 +82,67 @@ func (ghr *GHReleases) PublishProbe(driverVersion string, probePath string) erro
...
+		}
+		opt.Page = assetResponse.NextPage
+	}
+	return nil, fmt.Errorf("could not find matching asset for: %s", probeName)
+}
+
+// getReleaseByName uses the github API to identify the name of the release for the given driverVersion
+func (ghr *GHReleases) getReleaseByName(ctx context.Context, driverVersion string) (*github.RepositoryRelease, error) {
+	// Retrieve the releases
+	opt := &github.ListOptions{PerPage: 1}
```
Comment (VJftw) [inline pkg/repository/ghreleases/ghreleases.go:128]: nit: same here, don't forget to set this to `100` :)

**Review by VJftw — APPROVED** [review]
Awesome, nice one! 👍

**Pull request merged**